_source/_posts/2024-04-30-express-universal-logout.md

```diff
@@ -0,0 +1,674 @@
+
+```ts
+universalLogoutRoute.post('/global-token-revocation', async (req, res) => {
+  // 204 When the request is successful
+  const httpStatus = 204;
+
+  // 400 If the request is malformed
+  if (!req.body) {
+    res.status(400);
+  }
+  
+  // Make sure an API token is passed
+  const authHeaders = req.headers.authorization
+  if(!authHeaders){
+    return res.sendStatus(401)
+  }
+
+   // Find the user by email linked tothe org id associated with the API key provided
+  const domainOrgId = req['user']['id']
+  const newRequest:IRequestSchema = req.body;
+  const { email } = newRequest.sub_id;
+  const user = await prisma.user.findFirst({
+    where: {
+      email: email,
+      org: { id: domainOrgId } ,
+    },
+  });
+
+  // 404 User not found 
+  if (!user) {
+    res.sendStatus(404);
+  }
+
+universalLogoutRoute.use((err,req,res,next) => {
+  if(err){
+
+    return res.sendStatus(404)
+  }
+})
+```
+
+Moving right along, now that we have the target user of a specific org. Let's figure out how to target their application session and end it.
+
+## Implement the forced logout for a user
+
+In this section, we'll work towards ending a user's session and logging them out of the app.
+
+### End a user's session
+
+In Express, we'll need to access a user's session in the session store from the express-session library. Create a new file called sessionStore.ts under the `apps/api/src` folder and create a variable called store. A separate file to reference the `store` variable in multiple files. 
+
+```ts
+import {MemoryStore} from 'express-session';
+export const store = new MemoryStore()
+```
+
+For now, import `store` from the `session/Store` file at the top of the `apps/api/src/main.ts`. 
+
+```ts
+import { store } from './sessionsStore';
+```
+
+The top of the `apps/api/src/main.ts` file will now look like this:
+
+```ts
+import express from 'express';
+import { PrismaClient, Todo, User } from '@prisma/client';
+import passportLocal from 'passport-local';
+import passportOIDC from 'passport-openidconnect';
+import passport from 'passport';
+import session from 'express-session';
+import { universalLogoutRoute } from './universalLogout';
+import morgan from 'morgan';
+
+// Bearer Auth - Universal Logout
+import passportBearer from 'passport-http-bearer';
+import { store } from './sessionsStore';
+```
+
+Now add `store` within your session configuration:
+
+```ts
+app.use(session({
+ resave: false,
+ saveUninitialized: false,
+ secret: 'top secret',
+ cookie: {
+   http: false,
+   sameSite: 'lax'
+ },
+ store
+}));
+```
+
+Notice how the user session is configured in the `apps/api/src/main.ts` file via the Passport.js library's serialization and deserialization functions. We can instruct the Passport.js library to reference a user's session via user.id; this will come in handy soon. If you are curious, you can read more about how this works in the [Passport.js documentation](https://www.passportjs.org/concepts/authentication/sessions/). 
```

Comment:
```suggestion
Notice how the user session is configured in the `apps/api/src/main.ts` file via the Passport.js library's serialization and deserialization functions. We can instruct the Passport.js library to reference a user's session via `user.id`; this will come in handy soon. If you are curious, you can read more about how this works in the [Passport.js documentation](https://www.passportjs.org/concepts/authentication/sessions/). 
```